web-app/src/app/api/stripe/webhook/route.ts

```diff
@@ -0,0 +1,126 @@
+import { NextRequest, NextResponse } from "next/server";
+import Stripe from "stripe";
+
+import { getEnvSecrets } from "@/config/env.config";
+import prisma from "@/lib/db/prisma";
+import {
+  getFiatTransactionByServicePaymentId,
+  setFiatTransactionFailed,
+  setFiatTransactionSucceeded,
+} from "@/lib/db/services/fiatTransaction.service";
+import { FiatTransactionStatus } from "@/prisma/generated/client";
+
+const stripe = new Stripe(getEnvSecrets().STRIPE_SECRET_KEY);
+
+const updateFiatTransactionStatus = async (
+  session: Stripe.Checkout.Session,
+  status: "SUCCEEDED" | "FAILED",
+) => {
+  await prisma.$transaction(async (tx) => {
+    const fiatTransaction = await getFiatTransactionByServicePaymentId(
+      session.id,
+      tx,
+    );
+    if (!fiatTransaction) {
+      return NextResponse.json(
+        {
+          message: `Fiat transaction is not pending for session ${session.id}`,
+        },
+        { status: 400 },
+      );
+    }
+
+    if (session.client_reference_id !== fiatTransaction.id) {
+      return NextResponse.json(
+        {
+          message: `Session client reference id ${session.client_reference_id} does not match fiat transaction id ${fiatTransaction.id}`,
+        },
+        { status: 400 },
+      );
+    }
+
+    if (fiatTransaction.status !== FiatTransactionStatus.PENDING) {
+      return NextResponse.json(
+        { message: "Fiat transaction is not pending" },
+        { status: 400 },
+      );
+    }
+
+    switch (status) {
+      case "SUCCEEDED":
+        return await setFiatTransactionSucceeded(fiatTransaction, tx);
+      case "FAILED":
+        return await setFiatTransactionFailed(fiatTransaction, tx);
+    }
+  });
+};
+
+const handleSessionFailureEvents = async (session: Stripe.Checkout.Session) => {
+  console.info(`🔔  Payment failed for session ${session.id}`);
+  await updateFiatTransactionStatus(session, "FAILED");
+};
+
+const handleSessionSuccessEvents = async (session: Stripe.Checkout.Session) => {
+  const paymentStatus = session.payment_status;
+  if (paymentStatus !== "paid") {
+    return NextResponse.json(
+      { message: "Payment status is not paid" },
+      { status: 200 },
```

Comment:
Yes, this status is on purpose, because it's not a failure. It's just not yet paid.